src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -93,41 +93,89 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
             _packageSourceProvider = packageSourceProvider;
         }
 
-        public ImmutableArray<PackageSource> GetPackageSources()
+        public async ValueTask<ImmutableArray<PackageSource>?> TryGetPackageSourcesAsync(bool allowSwitchToMainThread, CancellationToken cancellationToken)
         {
-            // Only read from _packageSources once, since OnSourceProviderSourcesChanged could reset it to default at
-            // any time while this method is running.
-            var packageSources = _packageSources;
-            if (packageSources != null)
+            // Only read from _packageSourcesAsync once, since OnSourceProviderSourcesChanged could reset it to default
+            // at any time while this method is running.
+            var packageSourcesAsync = _packageSourcesAsync;
+            if (packageSourcesAsync is null)
             {
-                return packageSources;
+                lock (_gate)
+                {
+                    if (_packageSourcesAsync is null)
+                    {
+                        _packageSourcesAsync = ThreadingContext.JoinableTaskFactory.RunAsync(() => GetPackageSourcesImplAsync());
+                    }
+
+                    packageSourcesAsync = _packageSourcesAsync;
+                }
             }
 
-            try
+            if (packageSourcesAsync.IsCompleted)
             {
-                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
-                    .SelectAsArray(r => new PackageSource(r.Key, r.Value));
+                // Since the task is already completed, we know this 'await' will complete synchronously.
+                return await packageSourcesAsync;
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+            else if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                return await packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
```

Comment:
Isn't the original bug I blocked for back here?

https://github.com/dotnet/roslyn/pull/43718#discussion_r418415466

Because this now means a caller passing allowSwitchToMainThread = true can get cancelled due to the unexpected workspace change?

(I'm very much with @CyrusNajmabadi in that if this code path isn't active, let's just delete it.)

Reply:
I'm not sure if the intent was packageSourcesAsync now _never_ cancels due to workspace changes, but will now just retry until it finally suceeds, but then the _tokenSource.Token.CombineWith can probably go away?

Reply:
The `CombineWith` is no longer necessary here. Will remove.